CLI Demo Branch

abudgell opened the pull request:
Added a CLI, IMGUI integration, and put all as a ROS2 package cli_app with node cli_app_node.

abudgell committed:
Initial commit for CLI App

abudgell committed:
Rough draft demo of the CLIApp with launchable GUI

abudgell committed:
Split main.cpp into main.cpp, cli.cpp, and gui.cpp

abudgell committed:
Feat: Added ROS2 workspace for cli and implemented publisher for controller inputs

abudgell committed:
Feat: Added seperate file for a controller only imGUI

abudgell committed:
Changed loops to be 100HZ

Zaid-Duraid commented on .vscode/c_cpp_properties.json:
Should this file be in a .gitignore

Zaid-Duraid commented on .vscode/settings.json:
Should this file be in a .gitignore?

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/LICENSE:
What is this LICENSE file intended for? 

If it is for one of the libraries we're using, it might be worth it to indicate that in a README or put it in a [third-party-licenses.txt](https://softwareengineering.stackexchange.com/questions/234511/what-is-the-best-practice-for-arranging-third-party-library-licenses-paperwork) file.

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/archive/main.cpp:
Should this file be included?

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/eer_messages/CMakeLists.txt:
This seems to be a duplicate directory. It seems you put it here because you needed our custom eer message types...

This is more of a general comment, but would it be possible to move the cli_app to the `ROS2/ROS_workspace/src/` folder? That way it will be able to reference the custom message types in the existing `eer_messages`....

Moreover, soon we will likely make a new backend package for waterwitch (as soon as we fix a backend profile saving bug in `beaumont_pkg`). To keep consistent, we may be able to rename `beaumont_pkg` to `beaumont_backend` and name the new waterwitch package `waterwitch_backend`. To that effect,  we can name this package `cli_frontend`.

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/src/controller.cpp:
```diff
@@ -0,0 +1,254 @@
+        rclcpp::spin(node);
+    });
+
+    // Redirect stderr to /dev/null to suppress GStreamer warnings
```
Might be worth it to see if we can do a different approach to this (e.x. configure GStreamer to reduce verbosity).

This is fine for now though

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/src/controller.cpp:
```diff
@@ -0,0 +1,254 @@
+#endif
+
+    // Create window with graphics context
+    GLFWwindow* window = glfwCreateWindow(1920, 1080, "EER Controller Input", NULL, NULL);
```
Can we have the resolution be dynamically set based on screen resolution? I think this may be a simple fix

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/src/controller.cpp:
```diff
@@ -0,0 +1,254 @@
+    ImGui_ImplOpenGL3_Init(glsl_version);
+    
+
+    // Our state
```
I think this is some colour that's used in our GUI. However I'm not sure what the comment means

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/src/controller.cpp:
```diff
@@ -0,0 +1,254 @@
+        ImGui::NewFrame();
+
+
+        ImGui::Begin("Controller Box");
+        ImGui::Text("Use left stick on controller to move dot");
+
+        ImVec2 box_min = ImGui::GetCursorScreenPos();
+        ImVec2 box_max = ImVec2(box_min.x + 1000, box_min.y + 1000);
+        ImGui::InvisibleButton("box", ImVec2(1000, 1000));
+        ImDrawList* draw_list = ImGui::GetWindowDrawList();
+        draw_list->AddRect(box_min, box_max, IM_COL32(255, 255, 255, 255));
+
+        dot_position.x = std::max(box_min.x + dot_radius, std::min(dot_position.x, box_max.x - dot_radius));
+        dot_position.y = std::max(box_min.y + dot_radius, std::min(dot_position.y, box_max.y - dot_radius));
+
+        dot_position.x += input.sway / 127.0f;
+        dot_position.y += input.surge / 127.0f;
+
+        draw_list->AddCircleFilled(dot_position, dot_radius, IM_COL32(255, 0, 0, 255));
+
+        ImGui::End();
```
I like this. Helps to see if we have controller drift.

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/src/gui.cpp:
```diff
@@ -0,0 +1,414 @@
+}
+
+
+GLuint matToTexture(const cv::Mat &mat, GLuint textureID, GLenum minFilter, GLenum magFilter, GLenum wrapFilter)
+{
+    if (textureID == 0) {
+        glGenTextures(1, &textureID);
+        glBindTexture(GL_TEXTURE_2D, textureID);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, minFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, magFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, wrapFilter);
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, wrapFilter);
+        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, mat.cols, mat.rows, 0, GL_BGR, GL_UNSIGNED_BYTE, mat.ptr());
+    } else {
+        glBindTexture(GL_TEXTURE_2D, textureID);
+        glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, mat.cols, mat.rows, GL_BGR, GL_UNSIGNED_BYTE, mat.ptr());
+    }
+
+    return textureID;
+}
```
Might be worth adding a comment like "This function converts an OpenCV cv::Mat image to an OpenGL texture"

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/src/gui.cpp:
```diff
@@ -0,0 +1,414 @@
+                }
+            }
+
+            ImGui::Begin("Controller Input");
+            ImGui::Text("Axes:");
+            ImGui::Text("Surge: %d", input.surge);
+            ImGui::Text("Sway: %d", input.sway);
+            ImGui::Text("Heave: %d", input.heave);
+            ImGui::Text("Pitch: %d", input.pitch);
+            ImGui::Text("Yaw: %d", input.yaw);
+
+            ImGui::Text("Buttons:");
+            ImGui::Text("Open Claw: %s", input.open_claw ? "Pressed" : "Released");
+            ImGui::Text("Close Claw: %s", input.close_claw ? "Pressed" : "Released");
+            ImGui::Text("Heave Up: %s", input.heave_up ? "Pressed" : "Released");
+            ImGui::Text("Heave Down: %s", input.heave_down ? "Pressed" : "Released");
+            ImGui::Text("Pitch Up: %s", input.pitch_up ? "Pressed" : "Released");
+            ImGui::Text("Pitch Down: %s", input.pitch_down ? "Pressed" : "Released");
+            ImGui::Text("Brighten LED: %s", input.brighten_led ? "Pressed" : "Released");
+            ImGui::End();
```
There seems to be some duplicated functionality in this file with `controller.cpp` (with respect to the controller input publisher and associated code)

I understand that `controller.cpp` functionality is meant to be called when you provide the option "launch_controller" instead of "launch_gui". However, can we make it so that "launch_gui" itself references `controller.cpp` instead of duplicating it's functionality?

Zaid-Duraid commented on ROS2/ROS_workspace/cli_app/src/main.cpp:
```diff
@@ -0,0 +1,77 @@
+void printCLICommands();
+void handleCommand(const std::string& command);
+void printColoredAsciiImage();
+void printIntroduction();
```
Should we show the CLI options and commands after the introduction art?

abudgell replied:
This is done if you enter the command "help", enter a wrong command, or launch with "-h" as an argument

Zaid-Duraid reviewed (changes requested):
Looks great!

Two things to do before we merge this:
1. Adapt the openCV functionality to request camera URLs from our backend and show the associated MJPEG streams instead of webcam
2. See this review's comments

Zaid-Duraid committed:
Merge branch 'main' into CLI_test

abudgell committed:
Fix: Moved to common src file.
Fix: Delete second err_messages.
Feat: Added dynamic resolution detection for screen.
Feat: added ip camera support.

abudgell committed:
Moved everything to gui.cpp file in terms of controller vs gui as there was so much overlap between the two files.

abudgell committed:
Removed: LICENSE file for gui
